include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+   */
+  virtual ProxyAction onRequestMetadata() {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+
+  /**
+   * Call on a stream context to indicate that the request trailers have arrived.
+   */
+  virtual ProxyAction onResponseHeaders() {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+
+  /**
+   * Call on a stream context to indicate that body data has arrived.
+   * @param body_buffer_length the amount of data in the body buffer.
+   * @param end_of_stream is true if no more body data will be arriving.
+   */
+  virtual ProxyAction onResponseBody(int /* body_buffer_length */, bool /* end_of_stream */) {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+
+  /**
+   * Call on a stream context to indicate that the request trailers have arrived.
+   */
+  virtual ProxyAction onResponseTrailers() {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+
+  /**
+   * Call on a stream context to indicate that the request metadata has arrived.
+   */
+  virtual ProxyAction onResponseMetadata() {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+
+  // Root Context async events
+
+  /**
+   * Called on a Root Context when a response arrives for an outstanding httpCall().
+   * @param token is the token returned by the corresponding httpCall().
+   * @param headers are the number of headers in the response.
+   * @param body_size is the size in bytes of the response body.
+   * @param trailers is the number of trailres in the response.
+   */
+  virtual void onHttpCallResponse(uint32_t /* token */, uint32_t /* headers */,
+                                  uint32_t /* body_size */, uint32_t /* trailers */) {}
+
+  /**
+   * Called on a Root Context to indicate that an Inter-VM shared queue message has arrived.
+   * @token is the token returned by registerSharedQueue().
+   */
+  virtual void onQueueReady(uint32_t /* token */) { unimplemented(); }
+
+  // Context
+
+  /**
+   * Call when a stream has completed (both sides have closed) or on a Root Context when the VM is
+   * shutting down.
+   * @return true for stream contexts or for Root Context(s) if the VM can shutdown, false for Root
+   * Context(s) if the VM should wait until the Root Context calls the proxy_done() ABI call.  Note:
+   * the VM may (optionally) shutdown after some configured timeout even if the Root Context does
+   * not call proxy_done().
+   */
+  virtual bool onDone();
+
+  /**
+   * Call when a stream should log the final status of the stream. Occurs after onDone().
+   */
+  virtual void onLog();
+
+  /**
+   * Call when no further stream calls will occur.  This will cause the corresponding Context in the
+   * VM to be deleted.
+   */
+  virtual void onDelete();
+
+  /**
+   * Will be called on sever Wasm errors. Callees may report and handle the error (e.g. via an
+   * Exception) to prevent the proxy from crashing.
+   */
+  virtual void error(string_view message) {
+    std::cerr << message << "\n";
```

Comment:
Do you mean this to be here? Shouldn't this be platform specific?

Reply:
This is a placeholder which should be overridden.  It is overridden by Envoy with an Exception.